man/lc.1
.\" 
.\" lc manual page.
.\" (C) 2009 Carlo Kok
.\"  
.\" Author:
.\"   Carlo Kok (ck@remobjects.com)
.\"
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.TH Mono "lc"
.SH NAME
lc \- Mono License Compiler
.SH SYNOPSIS
.PP
.B lc -c filename -t targetassembly [-i reference...] [-v] [-o outputpath] [-nologo]
.SH DESCRIPTION
\fIlc\fP tool is a license compiler for Mono. It's used to convert a 
licenses.licx file to a resource that can be embedded into an executable. 
.PP
When using commercial components for .NET they often require a license
resource at runtime to make sure the developer was licensed to use this 
control. The most common way of making sure the component was properly
licensed is by using license resources. At compile time the developer
uses the \fIlc\fP tool to convert a licenses.licx file to a 
programname.exe.licenses file and embed that as a managed resource
during compilation.
.SH RUNTIME OPTIONS
The following options are available:
.TP
\fB--complist=filename\fR, \fB-c filename\fR
This option is used to specify the input file, the licenses.licx file
to create a license reesource for. This option is required. The license 
compiler will read each line in this file and try to load the types
specified on that line.
.TP
\fB--target=assemblyname\fR, \fB-t targetassembly\fR
full assembly name with file extension for the target assembly. The
license compiler uses this as a key when compiling the licenses. It
has to match the assembly the resource will be embedded in. This option
is required.
.TP
\fB--load=reference\fR, \fB-i reference\fR
Assembly reference to load before trying to compile a license. This
option is needed for assemblies where the references are not in the 
gac. Compilers should pass all references to the license compiler.
.TP
\fB--verbose,\fR, \fB-v\fR
Gives more verbose output about the license compiling process.
.TP
\fB--help\fR, \fB-h\fR
Displays usage instructions.
.TP
\fB--outdir=path\fR, \fB-o path\fR
Target path file the licenses resource. By default the license
compiler emits the resource in the current directory. The filename
will always match the target assembly passed to --target with .licenses 
added.
.TP
\fB--nologo\fR
Do not emit the "Mono License Compiler" logo  on stdout while compilng.

.SH LICENSE FILES
The format for the license files (licenses.licx) is the following:
.nf
	# comment
	namespace.type, assemblyname
	namespace.type, full-assemblyname
.fi
.Sp
All lines starting with a # are ignored. the first form requires
that the assembly was loaded through the --load,-i option but
does not require the full assembly name, only the first part.

The second form requires the full assembly name including locale,
version and publickeytoken.

.SH MAILING LISTS
Mailing lists are listed at the
http://www.mono-project.com/community/help/mailing-lists/
.SH WEB SITE
http://www.mono-project.com
.SH SEE ALSO
.PP
mono(1), mcs(1)



NEWS
This is a useless commit

Release notes for Mono are hosted on the web, please see:

	http://www.mono-project.com/docs/about-mono/releases/


mcs/class/Microsoft.Build.Tasks/Microsoft.Build.Tasks/LC.cs
//
// LC.cs: Task for license compiler
//
// Author:
//   Ankit Jain (jankit@novell.com)
//
// Copyright 2010 Novell, Inc (http://www.novell.com)
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
//
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHLCL THE AUTHORS OR COPYRIGHT HOLDERS BE
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DELCINGS IN THE SOFTWARE.


using System;
using System.Diagnostics;
using System.IO;
using Microsoft.Build.Framework;
using Microsoft.Build.Utilities;
using Mono.XBuild.Utilities;

namespace Microsoft.Build.Tasks {
	public class LC : ToolTaskExtension
	{
		public LC ()
		{
		}

		protected internal override void AddCommandLineCommands (
						CommandLineBuilderExtension commandLine)
		{
			if (Sources.Length == 0)
				return;

			foreach (ITaskItem item in Sources)
				commandLine.AppendSwitchIfNotNull ("--complist=", item.ItemSpec);

			commandLine.AppendSwitchIfNotNull ("--target=", LicenseTarget);

			if (ReferencedAssemblies != null)
				foreach (ITaskItem reference in ReferencedAssemblies)
					commandLine.AppendSwitchIfNotNull ("--load=", reference.ItemSpec);

			string outdir;
			if (Bag ["OutputDirectory"] != null)
				outdir = OutputDirectory;
			else
				outdir = ".";

			commandLine.AppendSwitchIfNotNull ("--outdir=", outdir);

			if (Bag ["NoLogo"] != null && NoLogo)
				commandLine.AppendSwitch ("--nologo");

			OutputLicense = new TaskItem (Path.Combine (OutputDirectory, LicenseTarget.ItemSpec + ".licenses"));
		}

		protected override string GenerateFullPathToTool ()
		{
			if (!string.IsNullOrEmpty (ToolPath))
				return Path.Combine (ToolPath, ToolExe);
			return ToolLocationHelper.GetPathToDotNetFrameworkFile (ToolExe, TargetDotNetFrameworkVersion.VersionLatest);
		}

		protected override bool ValidateParameters()
		{
			return Sources.Length > 0;
		}

		[Required]
		public ITaskItem LicenseTarget {
			get { return (ITaskItem) Bag ["LicenseTarget"]; }
			set { Bag ["LicenseTarget"] = value; }
		}

		public bool NoLogo {
			get { return GetBoolParameterWithDefault ("NoLogo", false); }
			set { Bag ["NoLogo"] = value; }
		}

		public string OutputDirectory {
			get { return (string) Bag ["OutputDirectory"]; }
			set { Bag ["OutputDirectory"] = value; }
		}

		[Output]
		public ITaskItem OutputLicense {
			get { return (ITaskItem) Bag ["OutputLicense"]; }
			set { Bag ["OutputLicense"] = value; }
		}

		public ITaskItem[] ReferencedAssemblies {
			get { return (ITaskItem[]) Bag ["ReferencedAssemblies"]; }
			set { Bag ["ReferencedAssemblies"] = value; }
		}

		[Required]
		public ITaskItem[] Sources {
			get { return (ITaskItem[]) Bag ["Sources"]; }
			set { Bag ["Sources"] = value; }
		}

		protected override string ToolName {
			get {
				return MSBuildUtils.RunningOnWindows ? "lc.bat" : "lc";
			}
		}
	}
}

